### ISSUE TITLE
Feature request: interpolating image sequence

### ISSUE THREAD

**Issue by styler00dollar** [ShuhongChen/eisai-anime-interpolator#4]
`interpolate.py` currently only takes 2 images as input. 
```
    ap.add_argument('img0', type=str)
    ap.add_argument('img1', type=str)
```
It would be nice if I could input an image sequence.

**Comment by ShuhongChen** [comment]
It's pretty complicated to explain timing specification across a sequence, so we figured it'd be easier to leave it at two frames

If you really wanted to do sequences programatically, you can try wrapping an image for loop around the [interpolate](https://github.com/ShuhongChen/eisai-anime-interpolator/blob/350d2418a0dd42784c19cf10dfdc420406718f00/_scripts/interpolate.py#L84) call, which actually does the work

**Comment by styler00dollar** [comment]
Managed to add this repo to [my code](https://github.com/styler00dollar/VSGAN-tensorrt-docker), so it can also be used for videos.

Random feedback to your code, do not use `_` as a variable [here](https://github.com/ShuhongChen/eisai-anime-interpolator/blob/350d2418a0dd42784c19cf10dfdc420406718f00/_scripts/interpolate.py#L48). I was quite confused why someone would name a variable like that. `_` usually means that it is not used.

**Comment by ShuhongChen** [comment]
> Managed to add this repo to my code

Cool repo; may I ask what your impression of eisai's speed/quality is compared to the others?  Not sure if you've run benchmarks yet

I see in your repo you have psnr-based deduplication; I remember we manually labeled a small dataset and found mean and max LAB diff more separable than mse-based

I'm not sure if I can dig up the exact result but here's some plots showing the separability by LAB features; the first plot shows the decision boundary (sorry for mess, it was rushed analysis)

Not sure if you're interested in this dedup method, but it's in our code [here](https://github.com/ShuhongChen/eisai-anime-interpolator/blob/350d2418a0dd42784c19cf10dfdc420406718f00/_util/frames_v0.py#L10) and [here](https://github.com/ShuhongChen/eisai-anime-interpolator/blob/350d2418a0dd42784c19cf10dfdc420406718f00/_scripts/duplicate_frame_features.py#L103)

![download](https://user-images.githubusercontent.com/11731345/179885510-780d6aeb-b9c1-4a3d-8b39-2c6bef8e7485.png)
![download](https://user-images.githubusercontent.com/11731345/179885524-e91d952b-41e1-4c6f-a1cc-d40137cdf69e.png)
![download](https://user-images.githubusercontent.com/11731345/179885530-1292f77b-885b-413f-be8b-96d4eb05652f.png)

> Random feedback to your code, do not use _ as a variable

I'm surprised you mention this instead of my use of `locals()`; I think I've done a lot of terrible things

**Comment by styler00dollar** [comment]
> I see in your repo you have psnr-based deduplication

Yeah I was very interested in your proposed dataset generation and I did not look at LAB yet. Didn't have time for it for now. Currently using PSNR and SSIM for now in my python API since that is easy to implement. A more complex approach I included is VMAF, which is used in the C++ rife extension.

> may I ask what your impression of eisai's speed/quality is compared to the others

I added a few benchmarks to my repo if you scroll down below. On my 1070ti I archive 3.5fps in 960x540 with eisai, which is quite slow in comparison to other architectures. As a comparison, with the same VapouSynth python API I reach (up to, depends on if ensembling and/or contextnet is used) 14fps in 1080p with rife4. Up to 34fps if I use a C++ ncnn extention to avoid CPU bottleneck. (Since this repo uses cupy, I have zero chance to get it working in ncnn. I even struggled to export your stuff as torch jit models, since you don't pass tensors.) So it is up to nearly 9.71x slower, while also having ~4x less data to work with. I am not saying that that speed needs to be reached, but that is quite a huge gap. Considering that TensorRT will have grid_sampler support soon, this gap will get bigger too, once I get this working for rife.

Since the resolution is kind of hardcoded to be 960x540, I had a hard time deciding how to benchmark it like i usually do, 480p, 720p and 1080p. I thought of modifying the code [here](https://github.com/ShuhongChen/eisai-anime-interpolator/blob/350d2418a0dd42784c19cf10dfdc420406718f00/_train/frame_interpolation/models/ssldtm.py), just replacing the resolution numbers, but I didn't test quality a lot if I do that yet.

I remember the paper talking about a major speedup, maybe I overlooked something, since 3.5fps with 2x scale is not that fast. I like the quality in normal scenes, but quite the major turnoff having random blobs appearing like I mentioned [here](https://github.com/ShuhongChen/eisai-anime-interpolator/issues/9). It happened quite a few times by watching normal animation, so I slowly tend to prefer rife again. I like it if it works at least, but I guess the models need finetuning. If I wouldn't encounter this artefact, I would say it is quite good in comparison to what I saw.

I only managed to train rife4 and cain until now, so it is a bit unfair to compare to film, ifunet, ifrnet, m2m, abme or enhanced sepconv on animation, since these used irl datasets.

> I'm surprised you mention this instead

I was trying to trim down around 5k lines of code, since I wanted one single file with the architecture. It was quite an adventure. I didn't notice everything. :D Your repository is gigantic.
![image](https://user-images.githubusercontent.com/51405565/180080317-90961986-b1f2-4c36-801f-28d804e21f3e.png)

**Comment by ShuhongChen** [comment]
> A more complex approach I included is VMAF

Cool, I hadn't heard of this one

> quite slow in comparison to other architectures

Thanks for the comparison; not surprising we're a lot slower

> Since this repo uses cupy

It's quite amazing there's no readily available gpu implementation of distance transform; I had to learn gpu programming just to do this for this project.  I picked cupy since hacking [softsplat](https://github.com/sniklaus/softmax-splatting) seemed simple enough, but I have no idea if it's a good library relative to others; if you have a more friendly framework please let me know

> Since the resolution is kind of hardcoded to be 960x540

Yes, it might be very hard to get things working for full hd; I'm not sure how I'd do it myself.  Our project is admittedly very "in vitro"

> I remember the paper talking about a major speedup

We compare in the supplementary to animeinterp, whose SGM flows take 2min each to calculate

> I slowly tend to prefer rife again

In my personal opinion, eisai model itself isn't that great; I think our repo+paper deliver two valuables:
1) RRLD pipeline: automated data collection is incredibly important; atd12k was a first step, but RRLD can actually scale up training.  As a university, releasing frames directly is pretty complicated, but we hope someone else can standardize a benchmark with RRLD; there are many issues with atd12k
2) LPIPS/CD: I really don't think psnr/ssim is appropriate; I hope we provided some formal arguments to convince people to move on from these for anime

> Your repository is gigantic

Most of it is dirty personal tools for research, I think at least half can actually be removed
